x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/process/IndexingStateProcessor.java

```diff
@@ -117,12 +138,62 @@ void persist(BytesReference bytes) throws IOException {
     }
 
     private static int findNextZeroByte(BytesReference bytesRef, int searchFrom, int splitFrom) {
-        for (int i = Math.max(searchFrom, splitFrom); i < bytesRef.length(); ++i) {
-            if (bytesRef.get(i) == 0) {
-                return i;
+        return bytesRef.indexOf((byte)0, Math.max(searchFrom, splitFrom));
+    }
+
+    @SuppressWarnings("unchecked")
+    static String extractDocId(BytesReference bytesRef) throws IOException {
+        String firstNonBlankLine = extractFirstNonBlankLine(bytesRef);
+        if (firstNonBlankLine == null) {
+            return null;
+        }
+        try (XContentParser parser =
+                 JsonXContent.jsonXContent.createParser(
+                     NamedXContentRegistry.EMPTY, LoggingDeprecationHandler.INSTANCE, firstNonBlankLine)) {
+            Map<String, Object> map = parser.map();
+            if ((map.get("index") instanceof Map) == false) {
+                throw new IllegalStateException("Could not extract \"index\" field out of [" + firstNonBlankLine + "]");
             }
+            map = (Map<String, Object>)map.get("index");
+            if ((map.get("_id") instanceof String) == false) {
+                throw new IllegalStateException("Could not extract \"index._id\" field out of [" + firstNonBlankLine + "]");
+            }
+            return (String)map.get("_id");
         }
-        return -1;
+    }
+
+    private static String extractFirstNonBlankLine(BytesReference bytesRef) {
+        for (int searchFrom = 0; searchFrom < bytesRef.length();) {
+            int newLineMarkerIndex = bytesRef.indexOf((byte) '\n', searchFrom);
+            int searchTo = newLineMarkerIndex != -1 ? newLineMarkerIndex : bytesRef.length();
+            String line = bytesRef.slice(searchFrom, searchTo - searchFrom).utf8ToString();
+            if (line.isBlank() == false) {
```

Comment:
Note that `isBlank` was first added in Java 11 so for the backport you'll need to do this differently.

Reply:
Yes, I'm pretty sure space is the only blank character that could occur (except for `\n` of course, but that's consumed by the line breaking).